### PR TITLE
Integrate EDGE Growth into Forecasting Model

### PR DESCRIPTION
**Pull request by mishtaiwi**
Brought in the EDGE growth functions under normits_demand/models/forecasting/edge_forecast.py
Call main function in normits_demand/models/forecasting/edge_replicant.py

Add the Cube export/file copying process and functions under normits_demand\utils being: c01_ExportCubeFunctions and r01_RunExportCube

### CODE REVIEW

**Comment by BenTaylor-TfN** [comment]
Will need to tidy this up with Matt's branch: https://github.com/Transport-for-the-North/NorMITs-Demand/tree/edge_replicant

Can discuss later

**Comment by BenTaylor-TfN** [comment]
I've made some initial changes to the code detailed below:

- Tided up imports - plenty of import weren't being used
- Formatted code with black - coding standards, easier to read
- Removed some `pd.options.mode.chained_assignment = None` lines
- Various bit of tidying
    - All this does is supress warnings. Deal with these warnings at the source instead.
- Moved files around
    - normits_demand/utils/c01_ExportCubeFunctions.py -> normits_demand/tools/edge_cube_extractor.py
        - Hopefully this is a tool at some point - just needs wrapping in a nice class / run-function
        - Also allows linking in with the rest of the EDGE replicant in future. That is, if the setup hasn't been done yet, just run the setup tool automatically.
    - normits_demand/utils/r01_RunExportCube.py run_tools/run_edge_cube_extractor.py
        - Links into the above point, making this a simple "setup and run" script.

**Comment by BenTaylor-TfN** [comment]
I'm now finished with my review/changes in the code for now. The code is now ready for a bit more of an in-depth review.

A couple of points from my review @mishtaiwi:
- variable/function naming should be all lower-case with underscores. There should never be an upper-case letter in a variable/function name.
    - Classes are a bit different. See the guide [here](https://google.github.io/styleguide/pyguide.html#3164-guidelines-derived-from-guidos-recommendations).
- Add type hints to functions 
- Run PyLint, MyPy (once type hints added), black, and PyDocStyle for general guidance on how to follow the standards.

Things for the next reviewer to keep an eye out for:
- Look for places that functionality can be replaced with framework functionality.
    - A good example of this is the `edge_cube_extractor.CheckFileExists`. There's already utils in the framework to do this.
- Note places code may need a refactor to fit with framework, even if not able to make changes now
- Keep an eye out for inconsistencies with the general flow of things in demand.
- Logging needs bringing into line with rest of demand so logging messages are picked up correctly.
    - I've chucked together an example in the cube extractor run/tool scripts

Happy to send over a copy of the coding standards if you want a look. Let's focus on getting 80% of the way there for now and hit all the main points. It doesn't have to be perfect if other work needs doing.

**Review by wsp-mbuckley — CHANGES_REQUESTED** [review]
Finished review most of the comments are fairly simple fixes and I think most aren't causing any errors in the process. One place where there is an error is on line 490 in `edge_forecast` where the internal flag isn't being set to 1 for internal zones.

Some other general points:
- Use `.loc` when setting values in DataFrames instead of just square brackets, not using `.loc` can sometimes cause errors
- The docstrings aren't following TfN's format, the first line in the docstring should have a brief explanation of the function / class with a more detailed explanation just below
- Reiterate Ben's points about type annotations, variable names and running pylint, mypy, black etc.

I think most of the comments should be fairly self-explanatory but we could have a chat about some of the areas where I've suggested using functionality already available in NorMITs demand.

File: normits_demand/tools/edge_cube_extractor.py [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+    zones_df = zones_df[["A", "STATIONCODE", "STATIONNAME"]]
+    # rename column
+    zones_df = zones_df.rename(columns={"A": "stn_zone_id"})
+    # remove '_' from station name and replace with ' '
+    zones_df["STATIONNAME"] = zones_df["STATIONNAME"].str.replace("_", " ")
+    # overwrite TLCs
+    for i, row in overwrite_tlcs.iterrows():
+        # get values
+        current_tlc = overwrite_tlcs.at[i, "NoRMS"]
+        overwrite_tlc = overwrite_tlcs.at[i, "Overwrite"]
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:146]: You can get the values directly from the row e.g. `row["NoRMS"]`

File: normits_demand/tools/edge_cube_extractor.py [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+    segment: str
+        segment of the overarching loop (e.g. purposes, periods, etc.)
+
+    Function
+    ----------
+    takes a cube .MAT file and export tabs to .csv files
+
+    Returns
+    -------
+    None.
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:182]: unneeded

File: normits_demand/tools/edge_cube_extractor.py [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+    """
+    # Export PT Demand
+    c_m = CUBEMatConverter(cube_exe)
+    c_m.mat_2_omx(in_mat, out_path, f"{out_csv}")
+    # open omx
+    omx_file = omx.open_file(pathlib.Path(out_path, f"{out_csv}.omx"))
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:188]: `OMXFile` should be able to handle this now

File: normits_demand/tools/edge_cube_extractor.py [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+            mx_df, pathlib.Path(out_path, f"{out_csv}_{single_mx}.csv", index=False)
+        )
+
+    # with omx_file.OMXFile(pathlib.Path(out_path, f"{out_csv}.omx")) as omx_mat:
+    #     for mx in omx_mat.matrix_levels:
+    #         mat = pd.DataFrame(omx_mat.get_matrix_level(mx),
+    #                            index=omx_mat.zones,
+    #                            columns=omx_mat.zones)
+    #         file_ops.write_df(mat, f"{out_path}/{out_csv}_{mx}.csv", index=False)
+
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:206]: This should now work with CUBE OMX files

File: normits_demand/tools/edge_cube_extractor.py [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+    output_folder : Path
+        full path top the folder where outputs to be saved.
+
+    Function
+    ----------
+    function produces PA From and To by time period
+
+    Returns
+    -------
+    None.
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:240]: remove

File: normits_demand/tools/edge_cube_extractor.py [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+
+    proc_single(
+        [
+            f'"{exe_cube}" "{output_folder}\\PT_Periods_FT.s" -Pvdmi /Start /Hide /HideScript',
+            f'del "{output_folder}\\*.prn"',
+            f'del "{output_folder}\\*.VAR"',
+            f'del "{output_folder}\\*.PRJ"',
+            f'del "{output_folder}\\PT_Periods_FT.s"',
+        ]
+    )
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:547]: the `subprocess.run` function can be used instead

File: normits_demand/tools/edge_cube_extractor.py [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+
+
+    Function
+    ----------
+    transposes the T matrices to get Od matrices by period and demand segmnet
+
+    Returns
+    -------
+    None.
+
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:569]: remvoe

File: normits_demand/tools/edge_cube_extractor.py [hunk 8]
```diff
@@ -0,0 +1,684 @@
...
+        "MW[97] = MW[97] + MW[107] ;OCA_F",
+        "MW[116] = MW[116] + MW[108] ;OCA_T",
+        "MW[100] = MW[100] + MW[109] ;ONCA",
+        "ENDRUN",
+    ]
+
+    with open(pathlib.Path(mats_folder, "PA2OD.s"), "w") as script:
+        script.write("\n".join(to_write))
+
+    proc_single(
```
Comment (wsp-mbuckley) [inline normits_demand/tools/edge_cube_extractor.py:676]: Use `subprocess.run`

File: run_models/run_forecast.py [hunk 9]
```diff
@@ -136,7 +136,7 @@ def main(
     if model in (forecast_cnfg.ForecastModel.TRIP_END, forecast_cnfg.ForecastModel.NTEM):
         tem_forecasting(params, model)
     elif model == forecast_cnfg.ForecastModel.EDGE:
-        edge_replicant.main(params)
+        edge_replicant.run_edge_growth(params)
```
Comment (wsp-mbuckley) [inline run_models/run_forecast.py:139]: Use `edge_replicant.run_edge_growth`

File: run_tools/run_edge_cube_extractor.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+"""Run Script for the EDGE CUBE extractor."""
+# ## IMPORTS ## #
+# Standard imports
+import os
+import sys
+import shutil
+
+# Third party imports
+from pathlib import Path
```
Comment (wsp-mbuckley) [inline run_tools/run_edge_cube_extractor.py:10]: group pathlib with built-in imports

File: run_tools/run_edge_cube_extractor.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+def run_extractor():
+    """Process Fixed objects."""
+    # time periods
+    periods = ["AM", "IP", "PM", "OP"]
+
+    # copy Cube files
+    for period in periods:
+        # read distance matrix
+        file_ops.check_file_exists(
+            Path(CUBE_CAT_RUN_PATH, f"Outputs/BaseAssign/{period}_stn2stn_costs.csv")
```
Comment (wsp-mbuckley) [inline run_tools/run_edge_cube_extractor.py:60]: You can use the `'/'` operator instead e.g. `CUBE_CAT_RUN_PATH / f"Outputs/BaseAssign/{period}_stn2stn_costs.csv"`

**Review by wsp-mbuckley — CHANGES_REQUESTED** [review]
Thanks for sorting out all the previous comments, just a few more mostly about return type annotations and docstring format we use the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).

There was an issue with the tool not running from `run_forecast` which should be fix with the following 2:
- Update the config class validation of `edge_factors_path`
- Line 139 in `run_forecast` needs updating to run the new function `run_edge_growth`

Then just test that it works with `run_forecast`

**Comment by BenTaylor-TfN** [comment]
Noting down some methodology questions/notes from my review. I want to make sure I understand why all the design decisions have been made.
@mishtaiwi - we may want a call to go through this in detail.
Let's add items to the forecasting backlog after discussion


### Questions
- Perhaps it makes sense to make an "EdgeReplicant" class / model to facilitate the following questions/changes
  - How much work would it be to allow the user to input a list of years and get the process to run for all years?
- Currently there are no outputs until the very end of the process (meaning that if it crashes before the end, we get nothing). 
  - Could we introduce a "Checkpointing" process where we output the matrices once they have been grown
  - Then we have a separate "output" step which aggregates the matrices to the output format
- Why does the cube extractor output 100+ matrices? Are there truly this many distinct growth factors? Can we aggregate these before input at all?
  - Could we introduce some sort of normalised naming convention that separate from NoRMS? Making it obvious which matrices are using which growth factors e.g. "hb_external_reduced_ticket...csv"
- `I:\NorMITs Demand\import\edge_replicant\apply_growth\Segments2UCs.csv` links input matrices to userclasses. Does it makes sense to add the growth method and internal / external to here instead of hardcoding it?
- It's currently a quite slow process. We could multiprocess it. Is there anything (memory / accessing the same data) you think would stop us from multiprocessing this?
- The process currently prints out base and future demand as it is running. I'm aware this is added to a log file during run, but it would be nice / cleaner to get this output as a singe log file

**Review by BenTaylor-TfN — APPROVED** [review]

**Pull request merged**